### PR TITLE
RM-9117   Upgrade to new logging infrastructure

### PR DESCRIPTION
**Pull request by KuJo7**
https://re-motion.atlassian.net/browse/RM-9117

### CODE REVIEW

File: Remotion/Core/Logging.Log4Net.UnitTests/Log4NetLoggerTests/DebugTest.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+namespace Remotion.UnitTests.Logging.Log4NetLogTests
+{
+  [TestFixture]
+  public class DebugTest : BaseTest
+  {
+    [Test]
+    public void IsEnabled_WithLevelAll ()
+    {
+      Logger.Repository.Threshold = Level.All;
+      Assert.That(Log.IsEnabled(Microsoft.Extensions.Logging.LogLevel.Debug), Is.True);
```
Comment (Frazerus) [inline Remotion/Core/Logging.Log4Net.UnitTests/Log4NetLoggerTests/DebugTest.cs:33]: We use the loglevel using everywhere, we should use it here as well.

Comment (Frazerus) [reply Remotion/Core/Logging.Log4Net.UnitTests/Log4NetLoggerTests/DebugTest.cs:33]: Please also do for the other tests in this project

File: Remotion/Core/Logging.Log4Net.UnitTests/Log4NetLoggerTests/DebugTest.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+// 
+using System;
+using log4net.Core;
+using Microsoft.Extensions.Logging;
+using NUnit.Framework;
+using Remotion.Logging;
+using Remotion.Logging.Log4Net.UnitTests.Log4NetLoggerTests;
+
+namespace Remotion.UnitTests.Logging.Log4NetLogTests
```
Comment (Frazerus) [inline Remotion/Core/Logging.Log4Net.UnitTests/Log4NetLoggerTests/DebugTest.cs:24]: Wrong namespace.

Comment (Frazerus) [reply Remotion/Core/Logging.Log4Net.UnitTests/Log4NetLoggerTests/DebugTest.cs:24]: Please fidx for all tests in this folder

File: Remotion/Core/Logging.Log4Net.UnitTests/Log4NetLoggerTests/FatalTest.cs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+using System;
+using log4net.Core;
+using NUnit.Framework;
+using Remotion.Logging;
+using Remotion.Logging.Log4Net.UnitTests.Log4NetLoggerTests;
+
+namespace Remotion.UnitTests.Logging.Log4NetLogTests
+{
+  [TestFixture]
+  public class FatalTest : BaseTest
```
Comment (Frazerus) [inline Remotion/Core/Logging.Log4Net.UnitTests/Log4NetLoggerTests/FatalTest.cs:26]: Why do we have these commented out?

File: Remotion/Core/Logging.Log4Net/Log4NetLogger.cs [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+using RemotionLoglevel = Remotion.Logging.LogLevel;
+
+namespace Remotion.Logging.Log4Net;
+
+/// <summary>
+/// Implements <see cref="Microsoft.Extensions.Logging.ILogger"/> to provide an implementation with log4net.
+/// </summary>
+public class Log4NetLogger : MicrosoftLogger
+{
+  public static Level Convert (MicrosoftLoglevel logLevel)
```
Comment (Frazerus) [inline Remotion/Core/Logging.Log4Net/Log4NetLogger.cs:33]: Depending on the answer for whether we want the extension methods with Microsoft.Logging or not these might be made private. 

If we want the extension methods with the old loglevels, we should convert in the extension methods. If we want the microsoft loglevels only, we can make the converts private and remove the unused one.

File: Remotion/Core/Logging.Log4Net/Log4NetLoggerProvider.cs [hunk 5]
```diff
@@ -15,25 +15,22 @@
...
-using log4net.Appender;
-using log4net.Config;
-using log4net.Core;
-using log4net.Layout;
-using log4net.Repository.Hierarchy;
-using Microsoft.Extensions.Logging;
-using Remotion.Reflection;
 using Remotion.Utilities;
 
+using MicrosoftLogger = Microsoft.Extensions.Logging.ILogger;
```
Comment (Frazerus) [inline Remotion/Core/Logging.Log4Net/Log4NetLoggerProvider.cs:20]: You can remove this using and use the normal one

File: Remotion/Core/UnitTests/Core.UnitTests.csproj [hunk 6]
```diff
@@ -19,6 +19,7 @@
   </ItemGroup>
   <ItemGroup>
     <ProjectReference Include="..\..\Development\UnitTesting.Compilation\Development.UnitTesting.Compilation.csproj" />
+    <ProjectReference Include="..\Extensions\Core.Extensions.csproj" />
```
Comment (Frazerus) [inline Remotion/Core/UnitTests/Core.UnitTests.csproj:22]:
```suggestion
```

**Review by Frazerus — CHANGES_REQUESTED** [review]

File: Remotion/Core/Core/Logging/LoggerExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,727 @@
...
+using Microsoft.Extensions.Logging;
+using Remotion.Utilities;
+
+using MicrosoftLogger = Microsoft.Extensions.Logging.ILogger;
+using MicrosoftLoglevel = Microsoft.Extensions.Logging.LogLevel;
+
+namespace Remotion.Logging
+{
+  /// <summary>
+  /// Provides extension methods used for loggerging.
```
[the remark this answers is not in the document]
Comment (Frazerus) [reply Remotion/Core/Core/Logging/LoggerExtensions.cs:29]: I think this was the only file I saw such errors.

**Review by MichaelKetting — CHANGES_REQUESTED** [review]
partial review, still need to check the tests and some prod-code. there's already some stuff to do, though.

File: Remotion/Core/Core/Logging/LoggerExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,727 @@
...
+
+    /// <summary>
+    /// Log a message object with the specified <paramref name="logLevel"/>.
+    /// </summary>
+    /// <param name="logger">The <see cref="MicrosoftLogger"/> instance where the message is to be logged.</param>
+    /// <param name="logLevel">The <see cref="MicrosoftLoglevel"/> of the message to be logged.</param>
+    /// <param name="message">The message object to log.</param>
+    public static void Log (this MicrosoftLogger logger, MicrosoftLoglevel logLevel, object? message)
+    {
+      Microsoft.Extensions.Logging.LoggerExtensions.Log(logger, logLevel, message?.ToString());
```
[the remark this answers is not in the document]
Comment (KuJo7) [reply Remotion/Core/Core/Logging/LoggerExtensions.cs:122]: Yes, this one is necessary!

[the remark this answers is not in the document]
Comment (MichaelKetting) [reply Remotion/Core/Core/Logging/LoggerExtensions.cs:122]: Ah, I see. The overload resolution matches the method on in this file with a parameter of type `object?` before matching the method with the parameter of type `string`.

File: Remotion/Core/Extensions.UnitTests/Logging/LoggerExtensionsObsoleteTests/DebugTest.cs
Comment (MichaelKetting) [inline Remotion/Core/Extensions.UnitTests/Logging/LoggerExtensionsObsoleteTests/DebugTest.cs:1]: I believe thare was a mistake about which files to use as basis. The intention was to use 
Remotion\Core\UnitTests\Logging\LogExtensionsTests
as basis, copy the folder and rename it as LogExtensionsObsoleteTests, then move to Core.Extensions.UnitTests project.

The current file looks like the LogExtensionsFolder from CoreExentsionsUnitTests was copied and named LogExtensionsObsoleteTests

Comment (MichaelKetting) [reply Remotion/Core/Extensions.UnitTests/Logging/LoggerExtensionsObsoleteTests/DebugTest.cs:1]: Update: there is a naming mismatch with the folders. I go ahead and fix the folder names.

**Review by MichaelKetting — CHANGES_REQUESTED** [review]

File: Remotion/Core/Core/Logging/LoggerExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,727 @@
...
+    /// </summary>
+    /// <param name="logger">The <see cref="MicrosoftLogger"/> instance where the message is to be logged.</param>
+    /// <param name="logLevel">The <see cref="MicrosoftLoglevel"/> of the message to be logged.</param>
+    /// <param name="eventID">The numeric identifier for the event.</param>
+    /// <param name="format">A string containing zero or more format items for the description of the context.</param>
+    /// <param name="args">An array containing zero or more objects to format.</param>
+    [StringFormatMethod("format")]
+    public static void LogFormat (this MicrosoftLogger logger, MicrosoftLoglevel logLevel, int eventID, string? format, params object?[] args)
+    {
+      logger.Log(logLevel, new EventId(eventID), null, format, args);
```
[the remark this answers is not in the document]
Comment (KuJo7) [reply Remotion/Core/Core/Logging/LoggerExtensions.cs:137]: yes, we do have one!

**Review by MichaelKetting — APPROVED** [review]
API is done based on the changes I made. What's left is to fix the failing tests. We can do this on the quest branch and rebase down after.

File: Remotion/Core/Logging.Log4Net/Log4NetLogger.cs (hunk 4 above)
Comment (MichaelKetting) [reply Remotion/Core/Logging.Log4Net/Log4NetLogger.cs]: Done.

File: Remotion/Core/Extensions.UnitTests/Logging/LoggerExtensionTests/LogTest.cs [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+// GNU Lesser General Public License for more details.
+// 
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+// 
+using System;
+using log4net.Core;
+using NUnit.Framework;
+using Remotion.Logging;
+using MicrosoftLoglevel = Microsoft.Extensions.Logging.LogLevel;
```
[the remark this answers is not in the document]
Comment (MichaelKetting) [reply Remotion/Core/Extensions.UnitTests/Logging/LoggerExtensionTests/LogTest.cs]: We stick with the Microsoft LogLevel in the new types, including the obsolete stuff. Most APIs used in practice don't directly expose the LogLevel anyway.

File: Remotion/Core/Extensions.UnitTests/Logging/LoggerExtensionsObsoleteTests/DebugTest.cs [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+// 
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+//
+using System;
+using log4net.Core;
+using NUnit.Framework;
+using Remotion.Logging;
+
+using MicrosoftLogLevel = Microsoft.Extensions.Logging.LogLevel;
```
[the remark this answers is not in the document]
Comment (MichaelKetting) [reply Remotion/Core/Extensions.UnitTests/Logging/LoggerExtensionsObsoleteTests/DebugTest.cs]: We stick with the Microsoft LogLevel in the new types, including the obsolete stuff. Most APIs used in practice don't directly expose the LogLevel anyway.

**Pull request merged**